docs/utilities.rst

```diff
@@ -18,8 +18,64 @@ Logging
     ti.set_logging_level(level)
 
 The default logging level is ``ti.INFO``.
-You can also override default logging level by setting the environment variable
-``TI_LOG_LEVEL`` to values such as ``trace`` and ``warn``.
+You can also override default logging level by setting the environment variable like
+``TI_LOG_LEVEL=warn``.
+
+
+.. _regression:
+Benchmarking and Regression Tests
+---------------------------------
+
+* Run ``ti benchmark`` to run tests in benchmark mode. This will record the performance of ``ti test``, and save it in ``benchmarks/output``.
+
+* Run ``ti regression`` to show the difference between previous result in ``benchmarks/baseline``. And you can see if the performance is increasing or decreasing after your commits. This is really helpful when your work is related to IR optimizations.
+
+* Run ``ti baseline`` to save the benchmark result to ``benchmarks/baseline`` for furture comparsion, this may be executed on performance related PRs, before they are merged into master.
+
+For example, this is part of the output by ``ti regression`` after enabling constant folding optimization pass:
+
+.. code-block::
+
+    linalg__________________polar_decomp______________________________
+    codegen_offloaded_tasks                       37 ->    39    +5.4%
+    codegen_statements                          3179 ->  3162    -0.5%
+    codegen_kernel_statements                   2819 ->  2788    -1.1%
+    codegen_evaluator_statements                   0 ->    14    +inf%
+
+    linalg__________________init_matrix_from_vectors__________________
+    codegen_offloaded_tasks                       37 ->    39    +5.4%
+    codegen_statements                          3180 ->  3163    -0.5%
+    codegen_kernel_statements                   2820 ->  2789    -1.1%
+    codegen_evaluator_statements                   0 ->    14    +inf%
+
+.. note::
+
+    Currently ``ti benchmark`` only support benchmarking number-of-statements, no time benchmarking is included since it depends on hardware performance and therefore hard to compare if the baseline is from another machine.
+    We are to purchase a fixed-performance machine as a time benchmark server at some point.
+    Discussion at: https://github.com/taichi-dev/taichi/issue/948
+
+
+The suggested workflow to do regression test for a PR in **reviewer-major** manner is:
```

Comment:
My feeling is that, for most of the times, the reviewers can simply ask the author to run and update the test baselines. So how about this:

```
For PRs that are deemed performance sensitive (e.g. an IR optimization), it's recommended that **the PR author** to run the regression tests, update and upload the baseline results.

The suggested workflow for **the PR author** to run the regression tests is:
...

**Reviewers** can also ask the PR author to run and upload the test result, if they consider the PR performance sensitive. If the reviewers would like to run the regression tests themselves, the suggested workflow is:

...
```